sympy/core/operations.py

```diff
@@ -504,3 +504,106 @@ def args(self):
     @staticmethod
     def _compare_pretty(a, b):
         return (str(a) > str(b)) - (str(a) < str(b))
+
+class AssocOpDispatcher(OpDispatcher):
+    """
+    Handler dispatcher for associative operators
+
+    Explanation
+    ===========
+
+    If a new class is meant to return the result other than `Add` or `Mul` when operated,
+    it should have its handler defined as class method. Also, if the arguments have the same
+    ``_op_priority`` the priority relation of it and other classes should be registered in
+    this dispatcher by ``register_priority`` method.
+
+    Examples
+    ========
+
+    >>> from sympy import Add, Expr, Symbol
+    >>> from sympy.core.operations import AssocOpDispatcher
+    >>> newadd = AssocOpDispatcher("newadd", Add)
+
+    >>> class NewExpr(Expr):
+    ...     @classmethod
+    ...     def _newadd_handler(cls, *args, **kwargs):
+    ...         return NewAdd(*args, **kwargs)
+    >>> class NewAdd(NewExpr, Add):
+    ...     pass
+
+    >>> @newadd.register_priority(Expr, NewExpr)
+    ... @newadd.register_priority(NewExpr, Expr)
+    ... @newadd.register_priority(NewExpr, NewExpr)
+    ... def _(_1, _2):
+    ...     if issubclass(_1, NewExpr):
+    ...         return _1
+    ...     else:
+    ...         return _2
+
+    >>> a, b = Symbol('a'), NewExpr()
+    >>> newadd(a, a) == Add(a, a)
+    True
+    >>> newadd(a, b) == NewAdd(a, b)
+    True
+    """
+
+    def register(self, *args, **kwargs):
+        raise NotImplementedError("To avoid confusion, use 'register_priority' method instead.")
+
+    def register_priority(self, cls1, cls2):
+        """
+        Register the priority between two types. Registered function must
+        return the type which has the handler that is to be used.
+        """
+        def _(func):
+            self.add((cls1, cls2), func)
+            return func
+        return _
+
+    def binary_select(self, cls1, cls2):
+        """
+        Select the class using registered priority. If the priority is
+        not defined, return None.
+        """
+        result = self.dispatch(cls1, cls2)
+        if result is None:
+            return None
+        else:
+            return result(cls1, cls2)
+
+    def find_handler(self, args):
+        """
+        Select the argument with highest priority, and return its handler.
+        """
+        args_data = []
+        for a in args:
+            # If _op_priority is not defined (e.g. `a` is int), consider its priority as 0
+            func, op_priority, handler = type(a), getattr(a, '_op_priority', 0), getattr(a, self.handlermethod, None)
+            if handler is not None:
```

Comment:
This loop is performance hotspot so it should be timed and optimised especially for the case where all handlers are None.

We don't need `type(a)` or `op_priority` if the handler is None so probably they should be computed inside the if. We can precompute `self.handlermethod` outside the loop. It might be possible to get a speedup using something like `operator.attrgetter` and `itertools` e.g.:
```
get = attrgetter(self.handlermethod)
zargs = zip(args, map(get, args))
fzargs = filter(itemgetter(1), zargs)
for a, handler in fzargs:
```
I don't know whether that would actually be faster but it's worth profiling/timing this loop for some case that should be fast like:
```
ints = [Integer(i) for i in range(10000)]
for n in [1, 10, 100, 1000, 10000]:
    intsn = ints[:n]
    add(intsn)   # <--- time this (timeit/cProfile)
```

Reply:
I tried your suggestion, but it didn't really affect the performance.